### ISSUE TITLE
don't see the code for confidence regularization

### ISSUE THREAD

**Issue by bhomass** [yueyu1030/COSINE#29]
when I check this part of the code in trainer.py

                loss = self.calc_loss(input = torch.log(softmax(logits)), \
                                        target= outputs_pseudo[0], \
                                        loss = self_training_loss, \
                                        thresh = self.args.self_training_eps, \
                                        soft = soft, \
                                        conf = 'entropy', \
                                        confreg = self.args.self_training_confreg)

                if self.args.self_training_contrastive_weight > 0:
                    contrastive_loss = self.contrastive_loss(input = torch.log(softmax(logits)), \
                                        feat = outputs_pseudo[-1], \
                                        target= outputs_pseudo[0], \
                                        conf = 'entropy', \
                                        thresh =  self.args.self_training_eps, \
                                        distmetric = self.args.distmetric, \
                                        )
                    loss = loss + self.args.self_training_contrastive_weight * contrastive_loss

                if self.args.gradient_accumulation_steps > 1:
                    loss = loss / self.args.gradient_accumulation_steps
                if torch.cuda.device_count() > 1:
                    loss = loss.mean()
                selftrain_loss += loss.item()

I see the classification loss added to the contrastive loss. Is the confidence regularization loss included in the self training?
where is it?

**Comment by yueyu1030** [comment]
It is in https://github.com/yueyu1030/COSINE/blob/main/trainer.py#L75.
This simple form of the confidence regularization term can be derived from the original equation in the paper.

**Comment by bhomass** [comment]
I checked the code against the formula a few times. it seems 

l -= confreg *( torch.sum(input * w.unsqueeze(1)) + np.log(n_classes_) * n_classes_ )

the n_classes_ at the end should be 1/n_classes_ if indeed ui = 1/C

and I couldn't account for the loss of 1/|C| factor for the first term  torch.sum(input * w.unsqueeze(1))

**Comment by yueyu1030** [comment]
I see. We omit the constant term in the code.

**Comment by bhomass** [comment]
is the * n_classes_ a mistake? should it be 1 / n_classes_?

**Comment by Janeey99** [comment]
I have same problem, but I still dont understand about R2 loss!

**Comment by bhomass** [comment]
R2 loss pushes the prediction confidence of every class to 1/number of classes, another word, no one class dominates over any other class. Essentially reduces any distinguishing capability of the classifier.

**Comment by yueyu1030** [comment]
R2 is only a regularization term for preventing over-confident pseudo labels. 

@bhomass The following comments are some kind suggestions for you. Before making comments on this repo, I would recommend you take some basic courses on machine learning and deep learning. Otherwise, I believe it is a waste of time to discuss this paper with you.

Some useful links: 
Machine Learning: https://bigml.com/ml101/
Deep Learning: https://www.coursera.org/specializations/deep-learning
Natural Language Processing: http://web.stanford.edu/class/cs224n/
